### PR TITLE
Threshold FHE Support

### PR DESCRIPTION
**Pull request by reneroliveira**
- Fixed EvalKeyMap pointers bug
- Add support for functions in Threshold FHE examples
- Exposing GetModulus, GetPlaintextModulus, and GetCyclotomicOrder as part of CryptoContext

### CODE REVIEW

File: src/lib/bindings.cpp [hunk 1]
```diff
@@ -446,6 +459,53 @@ void bind_crypto_context(py::module &m)
...
+             cc_MultiAddEvalMultKeys_docs,
+             py::arg("evalKey1"),
+             py::arg("evalKey2"),
+             py::arg("keyId") = "")
+        .def("MultiMultEvalKey", &CryptoContextImpl<DCRTPoly>::MultiMultEvalKey,
+             cc_MultiMultEvalKey_docs,
+             py::arg("privateKey"),
+             py::arg("evalKey"),
+             py::arg("keyId") = "")
+        //inline std::shared_ptr<std::map<usint, EvalKey<Element>>> MultiAddEvalSumKeys(const std::shared_ptr<std::map<usint, EvalKey<Element>>> evalKeyMap1, const std::shared_ptr<std::map<usint, EvalKey<Element>>> evalKeyMap2, const std::string &keyId = "")
```
Comment (mtriplett-duality) [inline src/lib/bindings.cpp:503]: Still required for experimentation?

**Review by mtriplett-duality — COMMENTED** [review]
Just style notes

File: src/lib/bindings.cpp [hunk 2]
```diff
@@ -104,6 +104,10 @@ void bind_crypto_context(py::module &m)
         //.def("GetScheme",&CryptoContextImpl<DCRTPoly>::GetScheme)
         //.def("GetCryptoParameters", &CryptoContextImpl<DCRTPoly>::GetCryptoParameters)
         .def("GetRingDimension", &CryptoContextImpl<DCRTPoly>::GetRingDimension, cc_GetRingDimension_docs)
+        .def("GetPlaintextModulus", &GetPlaintextModulusWrapper, cc_GetPlaintextModulus_docs)
+        .def("GetModulus", &GetModulusWrapper, cc_GetModulus_docs)
+        .def("GetCyclotomicOrder", &CryptoContextImpl<DCRTPoly>::GetCyclotomicOrder, cc_GetCyclotomicOrder_docs)
+        .def("GetModulus", &CryptoContextImpl<DCRTPoly>::GetModulus, cc_GetModulus_docs)
```
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/lib/bindings.cpp:110]: The threshold example prints out some parameters, the Modulos is one of them. But looking again at the code, I think it's wrong, the idea was to expose just the GetModulusWrapper, not the original one

File: src/include/docstrings/cryptocontext_docs.h [hunk 3]
```diff
@@ -875,6 +960,29 @@ const char* cc_Rescale_docs = R"pbdoc(
     :rtype: Ciphertext
 )pbdoc";
 
+const char* cc_RescaleInPlace_docs = R"pbdoc(
+    RescaleInPlace - An alias for OpenFHE ModReduceInPlace method. This is because ModReduceInPlace is called RescaleInPlace in CKKS.
```
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/include/docstrings/cryptocontext_docs.h:964]: I've copied this docstring from the main c++ library

File: src/include/docstrings/cryptocontext_docs.h [hunk 4]
```diff
@@ -59,6 +59,11 @@ const char* cc_GetRingDimension_docs = R"pbdoc(
     :rtype: int
 )pbdoc";
 
+// TODO (Oliveira, R.) - Add documentation
+const char* cc_GetPlaintextModulus_docs = "";
```
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/include/docstrings/cryptocontext_docs.h:63]: Yes, was intentionally blank, just to instantiate this string object in the main bindings.cpp file, without breaking the code. The intention was to complete the documentation together with other issue/milestone, but I can add it in a fix commit.

File: src/lib/bindings.cpp [hunk 5]
```diff
@@ -713,18 +793,28 @@ void bind_enums_and_constants(py::module &m)
...
+        .def("SetKeyTag", &PublicKeyImpl<DCRTPoly>::SetKeyTag);
+    py::class_<PrivateKeyImpl<DCRTPoly>, std::shared_ptr<PrivateKeyImpl<DCRTPoly>>>(m, "PrivateKey")
+        .def(py::init<>())
+        .def("GetKeyTag", &PrivateKeyImpl<DCRTPoly>::GetKeyTag)
+        .def("SetKeyTag", &PrivateKeyImpl<DCRTPoly>::SetKeyTag);
     py::class_<KeyPair<DCRTPoly>>(m, "KeyPair")
         .def_readwrite("publicKey", &KeyPair<DCRTPoly>::publicKey)
-        .def_readwrite("secretKey", &KeyPair<DCRTPoly>::secretKey);
+        .def_readwrite("secretKey", &KeyPair<DCRTPoly>::secretKey)
+        .def("good", &KeyPair<DCRTPoly>::good);
```
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/lib/bindings.cpp:812]: Added:

```c++
//KeyPair Docs
kp_good_docs = R"pbdoc(
    Checks whether both public key and secret key are non-null, or correctly initialized.

    :return: Result.
    :rtype: bool
)pbdoc";
```

File: src/include/docstrings/cryptocontext_docs.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/include/docstrings/cryptocontext_docs.h]: Added!

File: src/include/docstrings/cryptocontext_docs.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (reneroliveira) [reply src/include/docstrings/cryptocontext_docs.h:980]: I'll resolve the conversation, if we update the docs in the main library, I'll keep this one updated too.

File: src/lib/bindings.cpp [hunk 6]
```diff
@@ -501,10 +570,21 @@ void bind_crypto_context(py::module &m)
             "ClearEvalMultKeys", []()
             { CryptoContextImpl<DCRTPoly>::ClearEvalMultKeys(); },
             cc_ClearEvalMultKeys_docs)
+        .def_static(
+            "InsertEvalSumKey", &CryptoContextImpl<DCRTPoly>::InsertEvalSumKey,
+            cc_InsertEvalSumKey_docs,
+            py::arg("evalKeyMap"))
+        .def_static(
+            "InsertEvalMultKey", &CryptoContextImpl<DCRTPoly>::InsertEvalMultKey,
```
[the remark this answers is not in the document]
Comment (yspolyakov) [reply src/lib/bindings.cpp:575]: Not an issue  now

**Pull request merged**